mypy/server/aststripnew.py

```diff
@@ -0,0 +1,185 @@
+    """Reset a fine-grained incremental target to state before main pass of semantic analysis.
+
+    The most notable difference from the old version of strip_target() is that new semantic
+    analyzer doesn't have first pass where Vars and TypeInfos are pre-populated, so the stripping
+    is more thorough and straightforward.
+    """
+    visitor = NodeStripVisitor()
+    if isinstance(node, MypyFile):
+        visitor.strip_file_top_level(node)
+    else:
+        node.accept(visitor)
+
+
+class NodeStripVisitor(TraverserVisitor):
+    def __init__(self) -> None:
+        # The current active class.
+        self.type = None  # type: Optional[TypeInfo]
+        # This is True at class scope, but not in methods.
+        self.is_class_body = False
+        # By default, process function definitions. If False, don't -- this is used for
+        # processing module top levels.
+        self.recurse_into_functions = True
+
+    def strip_file_top_level(self, file_node: MypyFile) -> None:
+        """Strip a module top-level (don't recursive into functions)."""
+        self.recurse_into_functions = False
+        file_node.plugin_deps.clear()
+        file_node.accept(self)
+        file_node.names.clear()
+
+    def visit_block(self, b: Block) -> None:
+        if b.is_unreachable:
+            return
+        super().visit_block(b)
+
+    def visit_class_def(self, node: ClassDef) -> None:
+        """Strip class body and type info, but don't strip methods."""
+        # We need to delete any entries that were generated by plugins,
+        # since they will get regenerated.
+        to_delete = [v.node for v in node.info.names.values() if v.plugin_generated]
+        node.type_vars = []
+        node.base_type_exprs.extend(node.removed_base_type_exprs)
+        node.removed_base_type_exprs = []
+        node.defs.body = [s for s in node.defs.body
+                          if s not in to_delete]
+        with self.enter_class(node.info):
+            super().visit_class_def(node)
+        TypeState.reset_subtype_caches_for(node.info)
+        node.info = CLASSDEF_NO_INFO
+
+    def visit_func_def(self, node: FuncDef) -> None:
+        if not self.recurse_into_functions:
+            return
+        node.expanded = []
+        node.type = node.unanalyzed_type
+        # All nodes are non-final initially.
+        node.is_final = False
+        if node.type:
+            assert isinstance(node.type, CallableType)
+            node.type.variables = []
+        with self.enter_method(node.info) if node.info else nothing():
+            super().visit_func_def(node)
+
+    def visit_decorator(self, node: Decorator) -> None:
+        node.var.type = None
+        for expr in node.decorators:
+            expr.accept(self)
+        if self.recurse_into_functions:
+            # Only touch the final status if we re-process
+            # a method target.
+            node.var.is_final = False
```

Comment:
TBH, I just copied this from the old `aststrip`, but this indeed looks wrong, since decorators are processed as part of the module top-level. I will check if there is a test, and will add one if there is no test for this.